crates/sui-faucet/src/server.rs

```diff
@@ -35,12 +35,203 @@ use tracing::{info, warn};
 use uuid::Uuid;
 
 use crate::faucet::Faucet;
+use serde::Deserialize;
+use std::sync::Mutex;
+
+use anyhow::bail;
+use once_cell::sync::Lazy;
+
+/// Interval to cleanup expired tokens
+const CLEANUP_INTERVAL: u64 = 60; // 60 seconds
+/// Maximum number of requests per IP address
+const MAX_REQUESTS_PER_IP: u32 = 3;
+/// Interval to reset the request count for each IP address
+const RESET_TIME_INTERVAL_SECS: u64 = 12 * 3600; // 12 hours
+const CLOUDFLARE_URL: &str = "https://challenges.cloudflare.com/turnstile/v0/siteverify";
+const FAUCET_WEB_APP_URL: &str = "https://faucet.sui.io"; // make this lazy static env?
+
+static DISCORD_BOT: Lazy<Option<String>> = Lazy::new(|| std::env::var("DISCORD_BOT").ok());
+
+static TURNSTILE_SECRET_KEY: Lazy<Option<String>> =
+    Lazy::new(|| std::env::var("TURNSTILE_SECRET_KEY").ok());
+
+type IPAddr = String;
+
+/// Keep track of every IP address' requests.
+#[derive(Debug)]
+struct RequestsManager {
+    pub data: Mutex<BTreeMap<IPAddr, RequestInfo>>,
```

Comment:
Consider using an `Arc<DashMap<IPAddr, RequestInfo>>` (see [dashmap](https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html)) -- it is a lock-free concurrent hash map.